cosmology/StructureFormationManager.h

```diff
@@ -0,0 +1,597 @@
+                ParticlePositions.push_back(PosRow);
+                ParticleVelocities.push_back(VelRow);
+            }
+        }
+
+        // Create Particle container
+        size_type nloc = ParticlePositions.size(); 
+        std::cout << "rank: " << ippl::Comm->rank() << " Local number of particles: " << nloc << std::endl; 
+        std::shared_ptr<ParticleContainer_t> pc = this->pcontainer_m;
+        pc->create(nloc);  
+        pc->m = this->M_m/this->totalP_m;
+
+        auto R_host = pc->R.getHostMirror();
+        auto V_host = pc->V.getHostMirror();       
+        double a = this->a_m;
+        for (unsigned int i = 0; i < nloc; ++i) {
+            R_host(i)[0] = ParticlePositions[i][0];
+            R_host(i)[1] = ParticlePositions[i][1];
+            R_host(i)[2] = ParticlePositions[i][2];
+            V_host(i)[0] = ParticleVelocities[i][0]*pow(a, 1.5);
+            V_host(i)[1] = ParticleVelocities[i][1]*pow(a, 1.5);
+            V_host(i)[2] = ParticleVelocities[i][2]*pow(a, 1.5);
+        }
+
+        Kokkos::fence();
+        ippl::Comm->barrier();
+        Kokkos::deep_copy(pc->R.getView(), R_host);
+        Kokkos::deep_copy(pc->V.getView(), V_host);
+        Kokkos::fence();
+        ippl::Comm->barrier();
+        IpplTimings::stopTimer(ReadingTimer);
+
+        // Since the particles have moved spatially update them to correct processors
+        pc->update();
+
+        bool isFirstRepartition = false;
+        std::shared_ptr<FieldContainer_t> fc    = this->fcontainer_m;
+        if (this->loadbalancer_m->balance(this->totalP_m, this->it_m)) {
+                auto* mesh = &fc->getRho().get_mesh();
+                auto* FL = &fc->getFL();
+                this->loadbalancer_m->repartition(FL, mesh, isFirstRepartition);
+                printf("first repartition works \n");
+        }
+
+        
+        mes << "Assignment of positions and velocities done." << endl;
+
+    }
+
+    /**
+     * @brief Advance the simulation by one time step.
+     */
+    void advance() override {
+        if (this->stepMethod_m == "LeapFrog") {
+            LeapFrogStep();
+        }
+	else{
+            throw IpplException(TestName, "Step method is not set/recognized!");
+        }
+    }
+
+    /**
+     * @brief Perform a single LeapFrog step in the simulation.
+     */
+    void LeapFrogStep(){
+        // LeapFrog time stepping https://en.wikipedia.org/wiki/Leapfrog_integration
+
+        static IpplTimings::TimerRef VTimer           = IpplTimings::getTimer("pushVelocity");
+        static IpplTimings::TimerRef RTimer           = IpplTimings::getTimer("pushPosition");
+        static IpplTimings::TimerRef updateTimer      = IpplTimings::getTimer("update");
+        static IpplTimings::TimerRef domainDecomposition = IpplTimings::getTimer("loadBalance");
+        static IpplTimings::TimerRef SolveTimer       = IpplTimings::getTimer("solve");
+
+        //double dt                               = this->dt_m;
```

Comment:
is this commented line necessary? remove if not needed

Reply:
maybe add a comment explaining that the time step size is calculated according to what Blanca mentioned in her thesis:
"For the cosmological simulations, it was decided to adjust the timestep to the expansion of the universe. Instead of using a fixed ∆t, a fixed ∆ log a was implemented."